src/traversals/bfs.jl

```diff
@@ -58,46 +58,48 @@ bfs_tree(g::AbstractGraph, s::Integer; dir=:out) = tree(bfs_parents(g, s; dir=di
     gdistances!(g, source, dists)
 
 Fill `dists` with the geodesic distances of vertices in `g` from `source`.
-`dists` should be a vector of length `nv(g)`. Return `dists`.
-For vertices in disconnected components the default distance is -1.
+`dists` should be a vector of length `nv(g)` filled with `typemax(T)`. Return `dists`.
+For vertices in disconnected components the default distance is `typemax(T)`.
 """
-function gdistances!(g::AbstractGraph{T}, source, dists) where T
+function gdistances!(g::AbstractGraph{T}, source, vert_level) where T
     n = nv(g)
-    fill!(dists, typemax(T))
-    seen = zeros(Bool, n)
-    queue = Vector{T}(n)
-    @inbounds for i in 1:length(source)
-        queue[i] = source[i]
-        dists[source[i]] = 0
-        seen[source[i]] = true
+    visited = falses(n)
+    n_level = one(T)
+    cur_level = Vector{T}()
+    sizehint!(cur_level, n)
+    next_level = Vector{T}()
+    sizehint!(next_level, n)
+    @inbounds for s in source
+        vert_level[s] = zero(T)
+        visited[s] = true
+        push!(cur_level, s)
     end
-    head = 1
-    tail = length(source)
-    while head <= tail
-        current = queue[head]
-        distance = dists[current] + 1
-        head += 1
-        @inbounds for j in out_neighbors(g, current)
-            if !seen[j]
-                dists[j] = distance
-                tail += 1
-                queue[tail] = j
-                seen[j] = true
+    while !isempty(cur_level)
+        @inbounds for v in cur_level
+            @inbounds @simd for i in out_neighbors(g, v)
+                if !visited[i]
+                    push!(next_level, i)
+                    vert_level[i] = n_level
+                    visited[i] = true
+                end
             end
         end
+        n_level += one(T)
+        empty!(cur_level)
+        cur_level, next_level = next_level, cur_level
+        sort!(cur_level)
```

Comment:
Does this sort help performance, or is it for consistency of the output?

Reply:
It helps performance a good deal by maintaining cache locality. I did explicit tests.